diff -Naur base/game.c rev1/game.c
--- base/game.c
+++ rev1/game.c
@@ -31,7 +31,12 @@
 void game_describe(int8_t player, int8_t piece, uint8_t move) {
 	// Output stream
 	FILE * out = stdout;
-	fputc('\n', out);
+
+	// Show round linewise
+	if (player < 0)
+		fprintf(out, "\n%3d. ", board_moveCounter() / 2 + 1);
+	else 
+		fputc('\t', out);
 
 	// Long algebraic notation
 	if ((move & MOVE_CASTELING) != 0){
